File: components/tcHttptHandler.cpp
#include "tcHttptHandler.h"

#include "nsIHttpChannel.h"
#include "nsStringAPI.h"
#include "nsIURL.h"
#include "nsIStandardURL.h"
#include "nsISocketProviderService.h"
#include "nsISocketProvider.h"
#include "nsCOMPtr.h"
#include "nsXPCOM.h"
#include "nsIServiceManager.h"
#include "nsIGenericFactory.h"

NS_IMPL_ISUPPORTS3(tcHttptHandler,
                   nsIHttpProtocolHandler,
                   nsIProxiedProtocolHandler,
                   nsIProtocolHandler)

tcHttptHandler::tcHttptHandler()
{
    // httpHandler = do_GetService(NS_NETWORK_PROTOCOL_CONTRACTID_PREFIX "http");
    // NS_ASSERTION(httpHandler.get() != nsnull, "no http handler? from httpt");
    // return NS_OK;
}

nsresult
tcHttptHandler::Init()
{
    nsresult rv;
    nsIServiceManager *sm;

    rv = NS_GetServiceManager(&sm);
    if (NS_FAILED(rv)) return rv;

    rv = sm->GetServiceByContractID(NS_NETWORK_PROTOCOL_CONTRACTID_PREFIX "http",
                                    NS_GET_IID(nsIHttpProtocolHandler),
                                    getter_AddRefs(httpHandler));
    NS_ASSERTION(!NS_FAILED(rv), "getServiceByContractID failed");
    if (NS_FAILED(rv)) return rv;
    NS_ASSERTION(httpHandler.get() != nsnull, "no http handler? from httpt");
    return NS_OK;
}

NS_IMETHODIMP
tcHttptHandler::GetScheme(nsACString &aScheme)
{
    aScheme.AssignLiteral("httpt");
    return NS_OK;
}

NS_IMETHODIMP
tcHttptHandler::GetDefaultPort(PRInt32 *aPort)
{
    return httpHandler->GetDefaultPort(aPort);
}

NS_IMETHODIMP
tcHttptHandler::GetProtocolFlags(PRUint32 *aProtocolFlags)
{
    return httpHandler->GetProtocolFlags(aProtocolFlags);
}

NS_IMETHODIMP
tcHttptHandler::NewURI(const nsACString &aSpec,
                       const char *aOriginCharset,
                       nsIURI *aBaseURI,
                       nsIURI **_retval)
{
    return httpHandler->NewURI(aSpec, aOriginCharset, aBaseURI, _retval);
}

NS_IMETHODIMP
tcHttptHandler::NewChannel(nsIURI *aURI, nsIChannel **_retval)
{
    return httpHandler->NewChannel(aURI, _retval);
}

NS_IMETHODIMP
tcHttptHandler::AllowPort(PRInt32 aPort, const char *aScheme, PRBool *_retval)
{
    return httpHandler->AllowPort(aPort, aScheme, _retval);
}

NS_GENERIC_FACTORY_CONSTRUCTOR(tcHttptHandler);

static const nsModuleComponentInfo components[] = 
{
  {
    "tcHttptHandlerConstructor",
    TC_HTTPTHANDLER_CID,
    "@mozilla.org/network/protocol;1?name=httpt",
    tcHttptHandlerConstructor
  }
};

NS_IMPL_NSGETMODULE("tcHttptHandlerModule", components);

File: components/tcHttptHandler.js
const Cc = Components.classes;
const Ci = Components.interfaces;

Components.utils.import("resource://gre/modules/XPCOMUtils.jsm");

function tcHttptHandler() {
  dump("@@@@@@@@@@@@@@@@@@@@@\n");
}

tcHttptHandler.prototype = {
  classDescription: 'httpt protocol handler',
  classID: Components.ID('{45174059-8f44-45c6-9826-3f7e367cef2f}'),
  contractID: "@mozilla.org/network/protocol;1?name=httpt",
  QueryInterface: XPCOMUtils.generateQI([Ci.nsIProtocolHandler]),

  scheme: "httpt",
  defaultPort: 80,
  protocolFlags: Ci.nsIProtocolHandler.URI_NOAUTH,

  /* nsIHttpProtocolHandler */ 
  savedHttpHandler: 
    Components.classesByID["{4f47e42e-4d23-4dd3-bfda-eb29255e9ea3}"]. 
    getService(Components.interfaces.nsIHttpProtocolHandler), 

  allowPort: function(port, scheme) {
    return true;
  },

  newURI: function(aSpec, aOriginCharset, aBaseURI) {
    let url = Components.classes["@mozilla.org/network/standard-url;1"].
              createInstance(Components.interfaces.nsIStandardURL);
    url.init(Components.interfaces.nsIStandardURL.URLTYPE_AUTHORITY,
             this.defaultPort, aSpec, aOriginCharset, aBaseURI);
    return url.QueryInterface(Components.interfaces.nsIURI);
  },

  newChannel: function(aURI) {
    return this.savedHttpHandler.newProxiedChannel(aURI, null);
  }
};

const NSGetFactory = XPCOMUtils.generateNSGetFactory([tcHttptHandler]);